external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaDataConsumer.scala

```diff
@@ -516,13 +521,39 @@ private[kafka010] class KafkaDataConsumer(
     fetchedData.withNewPoll(records.listIterator, offsetAfterPoll)
   }
 
-  private def getOrRetrieveConsumer(): InternalKafkaConsumer = _consumer match {
-    case None =>
-      _consumer = Option(consumerPool.borrowObject(cacheKey, kafkaParams))
-      require(_consumer.isDefined, "borrowing consumer from pool must always succeed.")
-      _consumer.get
+  private[kafka010] def getOrRetrieveConsumer(): InternalKafkaConsumer = {
+    if (!_consumer.isDefined) {
+      retrieveConsumer()
+    }
+    if (!isConsumerUsingCurrentToken) {
+      logDebug("Cached consumer uses and old delegation token, invalidating.")
+      releaseConsumer()
+      consumerPool.invalidateKey(cacheKey)
+      fetchedDataPool.invalidate(cacheKey)
+      retrieveConsumer()
+    }
+    _consumer.get
+  }
+
+  private def retrieveConsumer(): Unit = {
+    _consumer = Option(consumerPool.borrowObject(cacheKey, kafkaParams))
+    require(_consumer.isDefined, "borrowing consumer from pool must always succeed.")
+  }
 
-    case Some(consumer) => consumer
+  private def isConsumerUsingCurrentToken: Boolean = {
```

Comment:
I'm still feeling that this method could be moved to KafkaTestUtil so that KafkaDataConsumer doesn't need to know about the details of how delegation token works. Either passing `_consumer.get` or `_consumer.get.kafkaParamsWithSecurity` as parameter of method.

But I would be OK if others are feeling it's OK to leave it as it is.

Reply:
I think it's a good idea, let me check...

Reply:
Moved.